src/main/java/net/logstash/logback/composite/loggingevent/KeyValuePairJsonProvider.java

```diff
@@ -0,0 +1,152 @@
+ * and {@link #excludeKvpKeyNames}:</p>
+ *
+ * <ol>
+ * <li>When {@link #includeKvpKeyNames} and {@link #excludeKvpKeyNames}
+ *     are both empty, then all entries will be included.</li>
+ * <li>When {@link #includeKvpKeyNames} is not empty and
+ *     {@link #excludeKvpKeyNames} is empty, then only those entries
+ *     with key names in {@link #includeKvpKeyNames} will be included.</li>
+ * <li>When {@link #includeKvpKeyNames} is empty and
+ *     {@link #excludeKvpKeyNames} is not empty, then all entries except those
+ *     with key names in {@link #excludeKvpKeyNames} will be included.</li>
+ * </ol>
+ *
+ * <p>It is a configuration error for both {@link #includeKvpKeyNames}
+ * and {@link #excludeKvpKeyNames} to be not empty.</p>
+ *
+ * <p>By default, for each entry in the KeyValuePair, the KeyValuePair key is output as the field name.
+ * This can be changed by specifying an explicit field name to use for an KeyValuePair key
+ * via {@link #addKvpKeyFieldName(String)}</p>
+ *
+ * <p>If the fieldName is set, then the properties will be written
+ * to that field as a subobject.
+ * Otherwise, the properties are written inline.</p>
+ */
+public class KeyValuePairJsonProvider extends AbstractFieldJsonProvider<ILoggingEvent> implements FieldNamesAware<LogstashFieldNames> {
+
+    /**
+     * See {@link KeyValuePairJsonProvider}.
+     */
+    private List<String> includeKvpKeyNames = new ArrayList<>();
+
+    /**
+     * See {@link KeyValuePairJsonProvider}.
+     */
+    private List<String> excludeKvpKeyNames = new ArrayList<>();
+
+    private final Map<String, String> kvpKeyFieldNames = new HashMap<>();
+
+    @Override
+    public void start() {
+        if (!this.includeKvpKeyNames.isEmpty() && !this.excludeKvpKeyNames.isEmpty()) {
+            addError("Both includeKvpKeyNames and excludeKvpKeyNames are not empty.  Only one is allowed to be not empty.");
+        }
+        super.start();
+    }
+
+    @Override
+    public void writeTo(JsonGenerator generator, ILoggingEvent event) throws IOException {
+        List<KeyValuePair> kvp = event.getKeyValuePairs();
+        if (kvp != null && !kvp.isEmpty()) {
+
+            boolean hasWrittenStart = false;
+
+            for (KeyValuePair kv : kvp) {
+                if (kv.key != null && kv.value != null
+                        && (includeKvpKeyNames.isEmpty() || includeKvpKeyNames.contains(kv.key))
+                        && (excludeKvpKeyNames.isEmpty() || !excludeKvpKeyNames.contains(kv.key))) {
+
+                    String fieldName = kvpKeyFieldNames.get(kv.key);
+                    if (fieldName == null) {
+                        fieldName = kv.key;
+                    }
+                    if (!hasWrittenStart && getFieldName() != null) {
+                        generator.writeObjectFieldStart(getFieldName());
+                        hasWrittenStart = true;
+                    }
+                    StructuredArguments.keyValue(fieldName, kv.value).writeTo(generator);
```

Comment:
This is the crux of the change, everything else is similar to the MdcJsonProvider.